### DISCUSSION TITLE
Limit solve time/iterations/... or cancel solving process

### DISCUSSION THREAD

**Discussion by SashmanGH** [coin-or/Sonnet#39] [Q&A] [answered]
Hi everyone,
I need to limit the solver duration. It would be great if I could let the solver run for a given duration like 5 minutes. If that is not possible, would it be possible to run the solver in background and then call a function to stop the solver? Or if even that won't work, could I tell the solver to do one iteration at a time so I could check the elapsed time inbetween iteration?

Thanks.
Sascha

**Comment by SashmanGH** [comment]
Hi,
thanks for the quick reply. Sorry, my description was to short. I use Sonnet/SonnetWrapper to call CBC from c#. Here is an example:
`     Sonnet.Model model = new Sonnet.Model("GoogleExample");
            
            Variable x = new Variable(VariableType.Integer);
            Variable y = new Variable(VariableType.Integer);

            model.Add(x + 7 * y <= 17.5);
            model.Add(0 <= x <= 3.5);
            model.Add(0 <= y);
            model.Objective = new Expression(x + 10 * y);

            
            Solver solver = new Solver(model);
            
            solver.Maximise();

`
From you linked example I can see a method setDblParam, but I can't find it in the .Net version. Am I missing something?

Thanks,
Sascha

**Reply by jhmgoossens** [reply, marked as answer]
Hi,
Apologies for the confusion. I didn’t realise your question was for Sonnet—I thought it was for cbc.. I deleted my original answer. 

For sonnet, the short answer is to either add the “sec” solver arg or cbcmodel setMaximumSeconds:

```
Solver solver = ….
OsiCbcSolverInterface osiCbc = solver.OsiSolver as OsiCbcSolverInterface;
osiCbc.AddCbcSolverArgs("-sec", "5");
```
or to use the setMaximumSeconds of CbcModel:
`osiCbc.getModelPtr().setMaximumSeconds(5.0;`
Which should work again since CBC #467 was closed. See [test/SonnetTest/Sonnet_CbcTests.cs](https://github.com/coin-or/Sonnet/blob/8a97583b15432b2cb8567b5cf0de5c8a264d73f3/test/SonnetTest/Sonnet_CbcTests.cs#L106)

For an example of how to use CBC Callbacks from sonnet, see also [Sonnet_CbcTests.cs](https://github.com/coin-or/Sonnet/blob/8a97583b15432b2cb8567b5cf0de5c8a264d73f3/test/SonnetTest/Sonnet_CbcTests.cs#L177)


For background: Sonnet relies on Osi, but Osi doesn’t include things like time limits. Therefore, these have to be done solver-specifically. 
Does this help?

**Comment by SashmanGH** [comment]
Hi,
thanks for your answer. Works as expredted. getModelPtr().setMaximumSeconds is still marked obsolete in my version so I used AddCbcSolverArgs. 
Now that I look at it, it seems to be a very basic question. So, thanks for you patience. 

Sascha
